x-pack/plugin/sql/qa/mixed-node/src/test/java/org/elasticsearch/xpack/sql/qa/mixed_node/SqlCompatIT.java

```diff
@@ -0,0 +1,127 @@
+            );
+            oldNodesClient = buildClient(
+                restClientSettings(),
+                nodes.getBWCNodes().stream().map(TestNode::getPublishAddress).toArray(HttpHost[]::new)
+            );
+        }
+    }
+
+    @AfterClass
+    public static void cleanUpClients() throws IOException {
+        IOUtils.close(newNodesClient, oldNodesClient, () -> {
+            newNodesClient = null;
+            oldNodesClient = null;
+            bwcVersion = null;
+        });
+    }
+
+    public void testNullsOrderBeforeMissingOrderSupport() throws IOException {
+        assumeTrue(
+            "expected some nodes without support for missing_order but got none",
+            bwcVersion.before(MISSING_ORDER_IN_COMPOSITE_AGGS_VERSION)
+        );
+
+        for (RestClient bwcClient : Arrays.asList(newNodesClient, oldNodesClient)) {
+            List<Integer> result = runOrderByNullsLastQuery(bwcClient);
+
+            assertEquals(3, result.size());
+            assertNull(result.get(0));
+            assertEquals(Integer.valueOf(1), result.get(1));
+            assertEquals(Integer.valueOf(2), result.get(2));
+        }
+    }
+
+    public void testNullsOrderWithMissingOrderSupport() throws IOException {
+        assumeTrue(
+            "expected all nodes with support for missing_order but got some without",
+            bwcVersion.onOrAfter(MISSING_ORDER_IN_COMPOSITE_AGGS_VERSION)
+        );
+
+        // TODO: add oldNodesClient once PR is backported to 7.x
+        for (RestClient bwcClient : Arrays.asList(newNodesClient)) {
+            List<Integer> result = runOrderByNullsLastQuery(bwcClient);
+
+            assertEquals(3, result.size());
+            assertEquals(Integer.valueOf(1), result.get(0));
+            assertEquals(Integer.valueOf(2), result.get(1));
+            assertNull(result.get(2));
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private List<Integer> runOrderByNullsLastQuery(RestClient queryClient) throws IOException {
+        Request putIndex = new Request("PUT", "/test");
+        putIndex.setJsonEntity("{\"settings\":{\"index\":{\"number_of_shards\":3}}}");
+        client().performRequest(putIndex);
+
+        Request indexDocs = new Request("POST", "/test/_bulk");
+        indexDocs.addParameter("refresh", "true");
+        StringBuilder bulk = new StringBuilder();
+        for (String doc : Arrays.asList("{\"int\":1,\"kw\":\"foo\"}", "{\"int\":2,\"kw\":\"bar\"}", "{\"kw\":\"bar\"}")) {
+            bulk.append("{\"index\":{}\n").append(doc).append("\n");
+        }
+        indexDocs.setJsonEntity(bulk.toString());
+        client().performRequest(indexDocs);
+
+        Request query = new Request("GET", "_sql");
+        query.setJsonEntity("{\"query\":\"SELECT int FROM test GROUP BY 1 ORDER BY 1 NULLS LAST\"}");
+        Response queryResponse = queryClient.performRequest(query);
```

Comment:
It would be nice to externalize these strings some how; once upgrading to Java11 declaring these strings should be less messy.